iree/hal/vulkan/serializing_command_queue.cc

```diff
@@ -39,199 +37,195 @@ namespace {
 // batch is ready to be submitted to GPU.
 // |wait_semaphores| and |signal_semaphores| will be filled with the binary
 // `VkSemaphores` on success.
-StatusOr<bool> TryToPrepareSemaphores(
+iree_status_t TryToPrepareSemaphores(
     const absl::InlinedVector<SemaphoreValue, 4>& batch_wait_semaphores,
     const absl::InlinedVector<SemaphoreValue, 4>& batch_signal_semaphores,
     const ref_ptr<TimePointFence>& batch_fence,
     absl::InlinedVector<VkSemaphore, 4>* wait_semaphores,
-    absl::InlinedVector<VkSemaphore, 4>* signal_semaphores) {
+    absl::InlinedVector<VkSemaphore, 4>* signal_semaphores,
+    bool* out_ready_to_submit) {
   IREE_TRACE_SCOPE0("TryToPrepareSemaphores");
-  IREE_DVLOG(3) << "TryToPrepareSemaphores";
+  *out_ready_to_submit = false;
 
   wait_semaphores->clear();
   for (const auto& timeline_semaphore : batch_wait_semaphores) {
-    IREE_DVLOG(3) << "Preparing binary VkSemaphore for timeline semaphore "
-                  << timeline_semaphore.semaphore << "..";
     // Query first to progress this timeline semaphore to the furthest.
-    IREE_ASSIGN_OR_RETURN(auto signaled_value,
-                          timeline_semaphore.semaphore->Query());
+    uint64_t signaled_value = 0;
+    IREE_RETURN_IF_ERROR(
+        iree_hal_semaphore_query(timeline_semaphore.first, &signaled_value));
 
     // If it's already signaled to a value greater than we require here,
     // we can just ignore this semaphore now.
-    if (signaled_value >= timeline_semaphore.value) {
-      IREE_DVLOG(3) << "..already signaled past; ignoring";
+    if (signaled_value >= timeline_semaphore.second) {
       continue;
     }
 
-    // SerializingCommandQueue only works with EmulatedTimelineSemaphore.
-    auto* emulated_semaphore =
-        static_cast<EmulatedTimelineSemaphore*>(timeline_semaphore.semaphore);
-
     // Otherwise try to get a binary semaphore for this time point so that
     // we can wait on.
-    VkSemaphore binary_semaphore = emulated_semaphore->GetWaitSemaphore(
-        timeline_semaphore.value, batch_fence);
-
-    if (binary_semaphore == VK_NULL_HANDLE) {
+    // TODO(antiagainst): if this fails we need to cancel.
```

Comment:
fyi @antiagainst , new TODOs in here with your name on them ;D

Reply:
hopefully we can let these remain as is and use this just to get us to the point when we can not need emulation in a shipping app #4494 - would feel much better about that :)

Reply:
Yep! The best code is no code :)